### PR TITLE
Fixing issue in data sync client

### PR DESCRIPTION
**Pull request by HiranyaKavishani**
## Purpose
This PR will fix the issues reported in https://github.com/wso2-extensions/identity-migration-resources/issues/123

### CODE REVIEW

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java [hunk 1]
```diff
@@ -206,26 +212,35 @@ private void addToStatementMap(Map<String, List<SQLStatement>> statements, SQLSt
...
+        String parentTableName = "";
+        String fkColumnName = "";
 
+        for (String tableName : syncTableList) {
             String schema = dataSourceManager.getSchema(tableName);
             String dataSourceType = dataSourceManager.getSourceDataSourceType(schema);
             DatabaseDialect databaseDialect = databaseDialectMap.get(dataSourceType);
             try (Connection sourceConnection = dataSourceManager.getSourceConnection(schema)) {
+                DatabaseMetaData metaData = sourceConnection.getMetaData();
+                ResultSet rs = metaData.getImportedKeys(sourceConnection.getCatalog(), null, tableName);
```
Comment (HiranyaKavishani) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java:237]: @omindu If this returns multiple, I think we have to create multiple triggers **L224**(**onDeleteTrigger = new Trigger(deleteTriggerName, parentTableName, targetTableName,
                            SYNC_OPERATION_DELETE, tableMetaData, SELECTION_POLICY_FOR_EACH_ROW, TRIGGER_TIMING_AFTER,
                            fkColumnName)**) for each of them. Could you please confirm?

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/Trigger.java [hunk 2]
```diff
@@ -44,6 +45,21 @@ public Trigger(String name, String sourceTableName, String targetTableName, Stri
...
+        this.sourceTableName = sourceTableName;
+        this.targetTableName = targetTableName;
+        this.triggerTiming = triggerTiming;
+        this.triggerEvent = triggerEvent;
+        this.tableMetaData = tableMetaData;
+        this.selectionPolicy = selectionPolicy;
+        this.foreignKey = foreignKey;
+    }
+
+
```
Comment (JKAUSHALYA) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/Trigger.java:62]: Remove these extra new lines.

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/impl/MySQLDatabaseDialect.java [hunk 3]
```diff
@@ -145,6 +145,37 @@ private static String getColumnEntryString(ColumnData columnEntry) {
...
+        // CREATE TRIGGER {triggerName} {triggerType} {triggerEvent} ON {sourceTableName} {selectionPolicy} BEGIN
+        // DELETE FROM {targetTableName} WHERE {foreignKeyColumnName}=columnValue; END;
+        String triggerStatement = String.format(SQL_TEMPLATE_DELETE_TRIGGER_MYSQL, triggerName, triggerType,
+                triggerEvent, sourceTableName, selectionPolicy, targetTableName, foreignKeyColumnName, columnValue);
+
+        sqlStatements.add(triggerStatement);
+        return sqlStatements;
+    }
+
+    //SQL_TEMPLATE_DELETE_TRIGGER_MYSQL
```
Comment (JKAUSHALYA) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/impl/MySQLDatabaseDialect.java:177]: Do we need this comment? If not, please remove it.

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java (hunk 1 above)
Comment (HiranyaKavishani) [reply components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java]: done the changes with fbd0759028559fb5303a7f2d0f6ced7c6aec2945

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/Trigger.java (hunk 2 above)
Comment (HiranyaKavishani) [reply components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/Trigger.java]: fixed with e03d2f1e6b6b1abc361cf6eac7d3da6ba7d0b6f4

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/impl/MySQLDatabaseDialect.java (hunk 3 above)
Comment (HiranyaKavishani) [reply components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/impl/MySQLDatabaseDialect.java]: Fixed with e03d2f1e6b6b1abc361cf6eac7d3da6ba7d0b6f4

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java [hunk 4]
```diff
@@ -206,26 +212,28 @@ private void addToStatementMap(Map<String, List<SQLStatement>> statements, SQLSt
...
         return scripts;
     }
 
-    public List<SQLStatement> generateTriggers() throws SyncClientException {
+    public List<SQLStatement> generateTriggers(Properties properties) throws SyncClientException {
 
         List<SQLStatement> sqlStatementList = new ArrayList<>();
-        for (String tableName : syncTableList) {
+        String parentTableName = "";
+        String fkColumnName = "";
```
Comment (omindu) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java:229]: These declarations can be moved inside the while loop.

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java [hunk 5]
```diff
@@ -260,6 +267,32 @@ private void addToStatementMap(Map<String, List<SQLStatement>> statements, SQLSt
                 addStatementsToStatementList(schema, SQL_STATEMENT_TYPE_SOURCE, sqlStatementList, onUpdateTriggerSQL);
                 addStatementsToStatementList(schema, SQL_STATEMENT_TYPE_SOURCE, sqlStatementList, onDeleteTriggerSQL);
 
+                Trigger onDeleteTriggersForChildTables;
+                DatabaseMetaData metaData = sourceConnection.getMetaData();
+                ResultSet rs = metaData.getImportedKeys(sourceConnection.getCatalog(), null, tableName);
```
Comment (omindu) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java:272]: Close the `ResultSet`

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/Trigger.java [hunk 6]
```diff
@@ -44,6 +45,20 @@ public Trigger(String name, String sourceTableName, String targetTableName, Stri
         this.selectionPolicy = selectionPolicy;
     }
 
+    public Trigger(String name, String sourceTableName, String targetTableName, String triggerEvent,
+                   TableMetaData tableMetaData,
+                   String selectionPolicy, String triggerTiming,String foreignKey) {
```
Comment (omindu) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/Trigger.java:50]:
```suggestion
                   String selectionPolicy, String triggerTiming, String foreignKey) {
```

Comment (omindu) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/Trigger.java:50]:
```suggestion
    public Trigger(String name, String sourceTableName, String targetTableName, String triggerEvent,
                   TableMetaData tableMetaData, String selectionPolicy, String triggerTiming, 
                   String foreignKey) {
```

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/internal/SyncClientComponent.java [hunk 7]
```diff
@@ -85,7 +85,7 @@ protected void activate(ComponentContext context) {
             if (prepareSync != null) {
                 syncService = new SyncService(properties);
                 String generateDDL = System.getProperty(JVM_PROPERTY_GENERATE_DDL);
-                syncService.generateScripts(generateDDL != null);
+                syncService.generateScripts(generateDDL != null, properties);;
```
Comment (omindu) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/internal/SyncClientComponent.java:88]: Why is it necessary to pass the `properties` here ? `SyncService` already has an instance of the system properties

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java [hunk 8]
```diff
@@ -206,26 +212,28 @@ private void addToStatementMap(Map<String, List<SQLStatement>> statements, SQLSt
...
+        String parentTableName = "";
+        String fkColumnName = "";
 
+        for (String tableName : syncTableList) {
             String schema = dataSourceManager.getSchema(tableName);
             String dataSourceType = dataSourceManager.getSourceDataSourceType(schema);
             DatabaseDialect databaseDialect = databaseDialectMap.get(dataSourceType);
             try (Connection sourceConnection = dataSourceManager.getSourceConnection(schema)) {
-
+                String syncTables = getProperty(JVM_PROPERTY_SYNC_TABLES, true, properties);
```
Comment (omindu) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java:236]: Why is it necessary to retrieve the `syncTables` from `properties` ? Is the items in `syncTableList` sufficient ?

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java [hunk 9]
```diff
@@ -381,4 +414,22 @@ private void addStatementsToStatementList(String schema, String statementType, L
 
         return databaseDialect.generateCreateTable(table);
     }
+
+    private String getProperty(String propertyName, boolean mandatory, Properties properties)
```
Comment (omindu) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java:418]: This method won't be necessary based on the comment at https://github.com/wso2-extensions/identity-migration-resources/pull/125/files#diff-e1344f082b677d743f1032b6161ddbb1R236

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java [hunk 10]
```diff
@@ -206,26 +212,28 @@ private void addToStatementMap(Map<String, List<SQLStatement>> statements, SQLSt
...
 
-        scripts.addAll(generateTriggers());
+        scripts.addAll(generateTriggers(properties));
         scripts.addAll(generateTables());
 
         return scripts;
     }
 
-    public List<SQLStatement> generateTriggers() throws SyncClientException {
+    public List<SQLStatement> generateTriggers(Properties properties) throws SyncClientException {
```
Comment (omindu) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java:225]: `Properties` object is not needed for this method based on https://github.com/wso2-extensions/identity-migration-resources/pull/125/files#diff-e1344f082b677d743f1032b6161ddbb1R236

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java [hunk 11]
```diff
@@ -64,6 +69,9 @@
 import static org.wso2.is.data.sync.system.util.Constant.DATA_SOURCE_TYPE_MYSQL;
 import static org.wso2.is.data.sync.system.util.Constant.DATA_SOURCE_TYPE_ORACLE;
 import static org.wso2.is.data.sync.system.util.Constant.DATA_SOURCE_TYPE_POSTGRESQL;
+import static org.wso2.is.data.sync.system.util.Constant.JVM_PROPERTY_SYNC_TABLES;
```
Comment (omindu) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java:72]: Unnecessary import based on https://github.com/wso2-extensions/identity-migration-resources/pull/125/files#r468158399

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/SyncService.java [hunk 12]
```diff
@@ -100,8 +100,8 @@ private void initiateDataTransformers() {
         dataTransformers.add(new AuthorizationCodeDataTransformerV580());
     }
 
-    public void generateScripts(boolean ddlOnly) throws SyncClientException {
+    public void generateScripts(boolean ddlOnly, Properties properties) throws SyncClientException {
```
Comment (omindu) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/SyncService.java:103]: Unnecessary introduction of `Properties`

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java [hunk 13]
```diff
@@ -260,6 +267,32 @@ private void addToStatementMap(Map<String, List<SQLStatement>> statements, SQLSt
                 addStatementsToStatementList(schema, SQL_STATEMENT_TYPE_SOURCE, sqlStatementList, onUpdateTriggerSQL);
                 addStatementsToStatementList(schema, SQL_STATEMENT_TYPE_SOURCE, sqlStatementList, onDeleteTriggerSQL);
 
+                Trigger onDeleteTriggersForChildTables;
```
Comment (omindu) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java:270]: Can be moved inside the while loop

Comment (HiranyaKavishani) [reply components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java]: Fixed with 54ad2a88f4dc379e72302caa9a8e87022d73a793

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/SyncService.java (hunk 12 above)
Comment (HiranyaKavishani) [reply components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/SyncService.java]: Fixed with 54ad2a88f4dc379e72302caa9a8e87022d73a793

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java (hunk 11 above)
Comment (HiranyaKavishani) [reply components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java]: Fixed with 54ad2a88f4dc379e72302caa9a8e87022d73a793

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java (hunk 10 above)
Comment (HiranyaKavishani) [reply components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java]: Fixed with 54ad2a88f4dc379e72302caa9a8e87022d73a793

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java (hunk 8 above)
Comment (HiranyaKavishani) [reply components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java]: Fixed with 54ad2a88f4dc379e72302caa9a8e87022d73a793

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/internal/SyncClientComponent.java (hunk 7 above)
Comment (HiranyaKavishani) [reply components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/internal/SyncClientComponent.java]: Fixed with 54ad2a88f4dc379e72302caa9a8e87022d73a793

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/Trigger.java (hunk 6 above)
Comment (HiranyaKavishani) [reply components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/Trigger.java]: Fixed with 54ad2a88f4dc379e72302caa9a8e87022d73a793

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java (hunk 5 above)
Comment (HiranyaKavishani) [reply components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java:267]: Fixed with 54ad2a88f4dc379e72302caa9a8e87022d73a793

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java (hunk 4 above)
Comment (HiranyaKavishani) [reply components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java]: Fixed with 54ad2a88f4dc379e72302caa9a8e87022d73a793

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java (hunk 9 above)
Comment (HiranyaKavishani) [reply components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java]: Fixed with d5a6fee8e8541053443fc392948e47e95319f6eb

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java [hunk 14]
```diff
@@ -260,6 +263,32 @@ private void addToStatementMap(Map<String, List<SQLStatement>> statements, SQLSt
...
 
+                DatabaseMetaData metaData = sourceConnection.getMetaData();
+                ResultSet rs = metaData.getImportedKeys(sourceConnection.getCatalog(), null, tableName);
+                int triggerCount = 0;
+                while (rs.next()) {
+                    String parentTableName = rs.getString(PK_TABLE_NAME);
+                    String fkColumnName = rs.getString(PK_COLUMN_NAME);
+                    String triggerName = getDeleteTriggerNameForChildTable(tableName, triggerCount);
+
+                    List<String> dropDeleteTriggerSQLForChildTable =
```
Comment (omindu) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java:274]: Take the `dropDeleteTriggerSQLForChildTable` and `onDeleteTriggerForChildTable` declaration outside the while loop and aggregate all the FK triggers to the lists.

Next move `addStatementsToStatementList` operations after the while loop.

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DatabaseDialect.java [hunk 15]
```diff
@@ -37,6 +37,15 @@
...
     List<String> generateCreateTrigger(Trigger trigger) throws SyncClientException;
 
+    /**
+     * Generate SQL statements for creating a trigger.
+     *
+     * @param trigger Trigger model containing trigger information.
+     * @return List of SQL statements related to trigger creation.
+     * @throws SyncClientException If error occurs while generating SQL statements.
+     */
+    List<String> generateDeleteTrigger(Trigger trigger) throws SyncClientException;
```
Comment (omindu) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DatabaseDialect.java:47]:
```suggestion
List<String> generateDeleteTrigger(Trigger trigger, Map<String, String> columnIds) throws SyncClientException {
```

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/impl/MySQLDatabaseDialect.java [hunk 16]
```diff
@@ -145,6 +145,35 @@ private static String getColumnEntryString(ColumnData columnEntry) {
...
+        String selectionPolicy = trigger.getSelectionPolicy();
+        String foreignKeyColumnName = trigger.getForeignKey();
+        TableMetaData tableMetaData = trigger.getTableMetaData();
+        List<ColumnData> columnDataList = tableMetaData.getColumnDataList();
+        String columnValue = "";
+
+        for (ColumnData columnEntry : columnDataList) {
+            if (SYNC_OPERATION_DELETE.equals(triggerEvent) && foreignKeyColumnName.equals(columnEntry.getName())) {
+                columnValue = "OLD." + columnEntry.getName();
+            }
```
Comment (omindu) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/impl/MySQLDatabaseDialect.java:165]: Should be changed to reflect the Map introduced with https://github.com/wso2-extensions/identity-migration-resources/pull/125/files#r469361615

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java [hunk 17]
```diff
@@ -260,6 +263,32 @@ private void addToStatementMap(Map<String, List<SQLStatement>> statements, SQLSt
                 addStatementsToStatementList(schema, SQL_STATEMENT_TYPE_SOURCE, sqlStatementList, onUpdateTriggerSQL);
                 addStatementsToStatementList(schema, SQL_STATEMENT_TYPE_SOURCE, sqlStatementList, onDeleteTriggerSQL);
 
+                DatabaseMetaData metaData = sourceConnection.getMetaData();
+                ResultSet rs = metaData.getImportedKeys(sourceConnection.getCatalog(), null, tableName);
+                int triggerCount = 0;
+                while (rs.next()) {
+                    String parentTableName = rs.getString(PK_TABLE_NAME);
+                    String fkColumnName = rs.getString(PK_COLUMN_NAME);
```
Comment (omindu) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java:271]: Capture the corresponding child table column name to address https://github.com/wso2-extensions/identity-migration-resources/pull/125/files#r469361615

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java [hunk 18]
```diff
@@ -260,6 +262,50 @@ private void addToStatementMap(Map<String, List<SQLStatement>> statements, SQLSt
...
+
+                            dropDeleteTriggerListSQLForChildTables.add(databaseDialect.generateDropTrigger(triggerName, targetTableName));
+
+                            Trigger onDeleteTriggersForChildTables = new Trigger(triggerName, parentTableName, targetTableName,
+                                    SYNC_OPERATION_DELETE, tableMetaData, SELECTION_POLICY_FOR_EACH_ROW, TRIGGER_TIMING_AFTER);
+                            onDeleteTriggerListForChildTables.add(databaseDialect.generateDeleteTrigger
+                                    (onDeleteTriggersForChildTables, columnIds));
+                        }
+                        triggerCount++;
+
```
Comment (omindu) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/dialect/DDLGenerator.java:292]:
```suggestion
```

File: components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/SQLQueryProvider.java [hunk 19]
```diff
@@ -67,8 +67,13 @@
             "(SYNC_ID INT)";
 
     public static final String SQL_TEMPLATE_CREATE_TRIGGER_MYSQL_KEY = "SQL_TEMPLATE_CREATE_TRIGGER_MYSQL";
+
     public static final String SQL_TEMPLATE_CREATE_TRIGGER_MYSQL = "CREATE TRIGGER %s %s %s ON %s %s " +
             "BEGIN INSERT INTO %s (%s) VALUES (%s); END";
+
+    public static final String SQL_TEMPLATE_DELETE_TRIGGER_MYSQL = "CREATE TRIGGER %s %s %s ON %s %s " +
+            "BEGIN DELETE FROM %s WHERE %s END";
```
Comment (omindu) [inline components/org.wso2.is.data.sync.client/src/main/java/org/wso2/is/data/sync/system/database/SQLQueryProvider.java:75]:
```suggestion
            "BEGIN DELETE FROM %s WHERE %s; END";
```

**Pull request merged**